File: labs/M15B-build-complete-agent/solution/coordinator.js
/**
 * M15B — Coordinator + Subagents (Solution — Node.js)
 * =====================================================
 *
 * Usage:
 *     node coordinator.js
 */

import Anthropic from "@anthropic-ai/sdk";
import { config } from "dotenv";

config();

const client = new Anthropic();
const MODEL = "claude-sonnet-4-20250514";

// Inline mock data (same 15 filings)
const MOCK_FILINGS = [
  { filing_number: "UCC-2024-NY-0012847", type: "UCC-1", state: "New York", status: "Active", debtor: { name: "Acme Corporation" }, secured_party: { name: "Atlantic Capital Partners" }, collateral_description: "All accounts receivable, inventory, equipment, and general intangibles now owned or hereafter acquired." },
  { filing_number: "UCC-2024-NY-0015921", type: "UCC-1", state: "New York", status: "Active", debtor: { name: "Acme Corporation" }, secured_party: { name: "Citibank N.A." }, collateral_description: "All deposit accounts, investment property, and letter-of-credit rights." },
  { filing_number: "UCC-2023-NY-0145678", type: "UCC-3", state: "New York", status: "Terminated", debtor: { name: "Harbor Shipping International Inc" }, secured_party: { name: "Citibank N.A." }, collateral_description: "TERMINATION." },
  { filing_number: "UCC-2024-NY-0019004", type: "UCC-1", state: "New York", status: "Active", debtor: { name: "Greenfield Logistics LLC" }, secured_party: { name: "JPMorgan Chase Bank N.A." }, collateral_description: "All inventory and accounts receivable." },
  { filing_number: "UCC-2024-CA-0098231", type: "UCC-1", state: "California", status: "Active", debtor: { name: "Pacific Ridge Technologies Inc" }, secured_party: { name: "Silicon Valley Bank" }, collateral_description: "All assets including IP, patents, trademarks, accounts." },
  { filing_number: "UCC-2024-CA-0101457", type: "UCC-1", state: "California", status: "Active", debtor: { name: "Acme Corporation" }, secured_party: { name: "Bank of America N.A." }, collateral_description: "All equipment and fixtures." },
  { filing_number: "UCC-2023-CA-0087652", type: "UCC-3", state: "California", status: "Amendment", debtor: { name: "Pacific Ridge Technologies Inc" }, secured_party: { name: "Silicon Valley Bank" }, collateral_description: "Amendment to add software and SaaS contracts." },
  { filing_number: "UCC-2023-TX-0187634", type: "UCC-1", state: "Texas", status: "Active", debtor: { name: "Lone Star Energy Solutions LP" }, secured_party: { name: "Wells Fargo Equipment Finance" }, collateral_description: "Specific equipment: Caterpillar excavators and Liebherr crane." },
  { filing_number: "UCC-2024-TX-0201337", type: "UCC-1", state: "Texas", status: "Active", debtor: { name: "Acme Corporation" }, secured_party: { name: "PNC Bank N.A." }, collateral_description: "All accounts receivable from Texas operations." },
  { filing_number: "UCC-2024-TX-0215890", type: "UCC-1", state: "Texas", status: "Active", debtor: { name: "Lone Star Energy Solutions LP" }, secured_party: { name: "Caterpillar Financial Services Corp" }, collateral_description: "Specific equipment: Caterpillar D10T2 tractors." },
  { filing_number: "UCC-2024-FL-0054219", type: "UCC-3", state: "Florida", status: "Amendment", debtor: { name: "Sunshine Medical Group PA" }, secured_party: { name: "TD Bank N.A." }, collateral_description: "Amendment to add MRI systems and CT scanner." },
  { filing_number: "UCC-2024-FL-0059811", type: "UCC-1", state: "Florida", status: "Active", debtor: { name: "Acme Corporation" }, secured_party: { name: "Atlantic Capital Partners" }, collateral_description: "All accounts receivable, inventory of Florida division." },
  { filing_number: "UCC-2024-IL-0076543", type: "UCC-1", state: "Illinois", status: "Active", debtor: { name: "Midwest Agricultural Cooperative" }, secured_party: { name: "Farm Credit Services of America" }, collateral_description: "All farm products, crops, livestock, equipment." },
  { filing_number: "UCC-2024-IL-0081290", type: "UCC-1", state: "Illinois", status: "Active", debtor: { name: "Acme Corporation" }, secured_party: { name: "JPMorgan Chase Bank N.A." }, collateral_description: "All assets of Illinois subsidiary." },
  { filing_number: "UCC-2023-IL-0069221", type: "UCC-3", state: "Illinois", status: "Amendment", debtor: { name: "Midwest Agricultural Cooperative" }, secured_party: { name: "Farm Credit Services of America" }, collateral_description: "Amendment to add grain equipment and John Deere combines." },
];

function searchFilings({ debtorName, state } = {}) {
  let r = MOCK_FILINGS;
  if (debtorName) r = r.filter(f => f.debtor.name.toLowerCase().includes(debtorName.toLowerCase()));
  if (state) r = r.filter(f => f.state.toLowerCase() === state.toLowerCase());
  return r;
}
function getFilingByNumber(num) { return MOCK_FILINGS.find(f => f.filing_number === num) || null; }
function calculateRisk(name) {
  const filings = searchFilings({ debtorName: name });
  if (!filings.length) return { debtor: name, risk_score: 0, risk_level: "UNKNOWN" };
  const active = filings.filter(f => f.status === "Active");
  const blanket = filings.filter(f => f.collateral_description.toLowerCase().includes("all a"));
  const amend = filings.filter(f => f.type === "UCC-3");
  const states = [...new Set(filings.map(f => f.state))];
  const parties = [...new Set(filings.map(f => f.secured_party.name))];
  const score = Math.min(1.0, Math.round((active.length * 0.15 + blanket.length * 0.2 + amend.length * 0.05 + (states.length > 1 ? 0.1 : 0) + (parties.length > 1 ? 0.1 : 0)) * 100) / 100);
  return { debtor: name, risk_score: score, risk_level: score >= 0.7 ? "HIGH" : score >= 0.4 ? "MEDIUM" : "LOW", total_filings: filings.length, active_filings: active.length };
}

const SEARCH_TOOLS = [
  { name: "search_filings", description: "Search UCC filings.", input_schema: { type: "object", properties: { debtor_name: { type: "string" }, state: { type: "string" } }, required: [] } },
  { name: "get_filing_details", description: "Get filing details.", input_schema: { type: "object", properties: { filing_number: { type: "string" } }, required: ["filing_number"] } },
];
const RISK_TOOLS = [
  { name: "calculate_risk_score", description: "Calculate risk.", input_schema: { type: "object", properties: { debtor_name: { type: "string" } }, required: ["debtor_name"] } },
  { name: "search_filings", description: "Search filings for context.", input_schema: { type: "object", properties: { debtor_name: { type: "string" }, state: { type: "string" } }, required: [] } },
];

function executeTool(name, input) {
  if (name === "search_filings") { const r = searchFilings({ debtorName: input.debtor_name, state: input.state }); return !r.length ? JSON.stringify({ message: "No filings found" }) : JSON.stringify(r.map(f => ({ filing_number: f.filing_number, debtor: f.debtor.name, secured_party: f.secured_party.name, state: f.state, status: f.status, collateral: f.collateral_description.slice(0,120)+"..." })), null, 2); }
  if (name === "get_filing_details") { const f = getFilingByNumber(input.filing_number); return f ? JSON.stringify(f, null, 2) : JSON.stringify({ error: "Not found" }); }
  if (name === "calculate_risk_score") return JSON.stringify(calculateRisk(input.debtor_name), null, 2);
  return JSON.stringify({ error: `Unknown: ${name}` });
}

function observe(l, m) { console.log(`\n${"=".repeat(60)}\n[${l}] ${m}\n${"=".repeat(60)}`); }
function observeAgent(n, a) { console.log(`\n${"─".repeat(60)}\n[AGENT: ${n}] ${a}\n${"─".repeat(60)}`); }
function observeHandoff(f, t, s) { console.log(`\n${"─".repeat(60)}\n[HANDOFF] ${f} → ${t} (${s} chars)\n${"─".repeat(60)}`); }

async function runSubagent(name, system, task, tools, maxTurns = 6) {
  observeAgent(name, `Starting: ${task.slice(0, 100)}...`);
  const messages = [{ role: "user", content: task }];
  for (let t = 0; t < maxTurns; t++) {
    const resp = await client.messages.create({ model: MODEL, max_tokens: 4096, system, tools, messages });
    if (resp.stop_reason !== "tool_use") { const text = resp.content.filter(b => b.type === "text").map(b => b.text).join(""); observeAgent(name, `Complete (${text.length} chars)`); return text; }
    const results = [];
    for (const b of resp.content) { if (b.type === "tool_use") results.push({ type: "tool_result", tool_use_id: b.id, content: executeTool(b.name, b.input) }); }
    messages.push({ role: "assistant", content: resp.content });
    messages.push({ role: "user", content: results });
  }
  return `${name} did not complete.`;
}

class Coordinator {
  constructor() { this.history = []; }

  async run(query) {
    observe("COORDINATOR", `Received: ${query}`);
    this.history.push({ role: "user", content: query });
    const histCtx = this.history.slice(-10).map(e => `**${e.role === "user" ? "User" : "Assistant"}**: ${e.content.slice(0,200)}`).join("\n");
    const q = query.toLowerCase();
    const needsRisk = ["risk", "exposure", "assess", "evaluate"].some(w => q.includes(w));
    const needsSearch = ["find", "search", "filing", "what about"].some(w => q.includes(w));
    const taskCtx = `${query}\n\n## Previous Conversation\n${histCtx}`;

    let filingResults = null, riskResults = null;
    if (!needsRisk || (needsSearch && needsRisk) || (!needsSearch && !needsRisk)) {
      filingResults = await runSubagent("Filing Search", "You are a UCC filing search specialist. Search and retrieve filings. Be thorough.", taskCtx, SEARCH_TOOLS);
      observeHandoff("Filing Search", "Coordinator", filingResults.length);
    }
    if (needsRisk || (!needsSearch && !needsRisk)) {
      const riskTask = filingResults ? `${taskCtx}\n\n## Filing Data\n${filingResults}` : taskCtx;
      riskResults = await runSubagent("Risk Analysis", "You are a risk analysis specialist. Calculate and explain lien risk.", riskTask, RISK_TOOLS);
      observeHandoff("Risk Analysis", "Coordinator", riskResults.length);
    }

    const parts = [`User query: ${query}`];
    if (filingResults) parts.push(`## Filing Results\n${filingResults}`);
    if (riskResults) parts.push(`## Risk Results\n${riskResults}`);
    const synth = await client.messages.create({ model: MODEL, max_tokens: 4096, system: "Synthesize UCC research results. Cite filing numbers.", messages: [{ role: "user", content: parts.join("\n\n") }] });
    const final = synth.content.filter(b => b.type === "text").map(b => b.text).join("");
    this.history.push({ role: "assistant", content: final.slice(0, 500) });
    observe("COORDINATOR", "Complete");
    return final;
  }
}

console.log("=".repeat(60));
console.log("M15B — Coordinator (SOLUTION — Node.js)");
console.log("=".repeat(60));

const coord = new Coordinator();
const r1 = await coord.run("Find all UCC filings for Acme Corporation in New York");
console.log(`\nANSWER:\n${r1}`);
const r2 = await coord.run("What's the overall risk level for Acme Corporation?");
console.log(`\nANSWER:\n${r2}`);
const r3 = await coord.run("What about their filings in Texas?");
console.log(`\nANSWER:\n${r3}`);

File: labs/M15B-build-complete-agent/solution/coordinator.py
"""
M15B — Coordinator + Subagents (Solution)
===========================================
Complete coordinator with 2 specialist subagents and conversation memory.

Usage:
    python coordinator.py
"""

import json
import os
from dotenv import load_dotenv

load_dotenv()

import anthropic
from config import MODEL, MAX_SUBAGENT_TURNS, MAX_HISTORY_TURNS
from tools import TOOL_DEFINITIONS, execute_tool

client = anthropic.Anthropic()


# =============================================================================
# OBSERVATION HELPERS
# =============================================================================

def observe(label: str, message: str) -> None:
    print(f"\n{'=' * 60}")
    print(f"[{label}] {message}")
    print(f"{'=' * 60}")


def observe_agent(name: str, action: str) -> None:
    print(f"\n{'─' * 60}")
    print(f"[AGENT: {name}] {action}")
    print(f"{'─' * 60}")


def observe_handoff(from_a: str, to_a: str, size: int) -> None:
    print(f"\n{'─' * 60}")
    print(f"[HANDOFF] {from_a} → {to_a} ({size} chars)")
    print(f"{'─' * 60}")


# =============================================================================
# SUBAGENT TOOL SETS
# =============================================================================

FILING_SEARCH_TOOLS = [t for t in TOOL_DEFINITIONS if t["name"] in ("search_filings", "get_filing_details")]
RISK_ANALYSIS_TOOLS = [t for t in TOOL_DEFINITIONS if t["name"] in ("calculate_risk_score", "search_filings")]


# =============================================================================
# SUBAGENT RUNNER
# =============================================================================

def run_subagent(agent_name: str, system_prompt: str, task: str,
                 tools: list, max_turns: int = None) -> str:
    """Run a specialist subagent with isolated context."""
    if max_turns is None:
        max_turns = MAX_SUBAGENT_TURNS

    observe_agent(agent_name, f"Starting: {task[:100]}...")
    messages = [{"role": "user", "content": task}]

    try:
        for turn in range(max_turns):
            response = client.messages.create(
                model=MODEL,
                max_tokens=4096,
                system=system_prompt,
                tools=tools,
                messages=messages,
            )

            if response.stop_reason != "tool_use":
                text = ""
                for block in response.content:
                    if hasattr(block, "text"):
                        text += block.text
                observe_agent(agent_name, f"Complete ({len(text)} chars)")
                return text

            tool_results = []
            for block in response.content:
                if block.type == "tool_use":
                    result = execute_tool(block.name, block.input)
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": result,
                    })

            messages.append({"role": "assistant", "content": response.content})
            messages.append({"role": "user", "content": tool_results})

        return f"{agent_name} did not complete within {max_turns} turns."

    except Exception as e:
        observe_agent(agent_name, f"Error: {str(e)}")
        return json.dumps({"error": f"{agent_name} failed: {str(e)}"})


# =============================================================================
# SPECIALIST SUBAGENTS
# =============================================================================

def run_filing_search(task: str) -> str:
    """Filing Search Subagent."""
    system_prompt = """You are a UCC filing search specialist. Your ONLY job is to search for
and retrieve UCC filing records. Use search_filings to find filings by debtor name
and/or state, then use get_filing_details for full information.

Report your findings as structured data: list each filing with its number, debtor,
secured party, state, status, type, filing date, and collateral description.
Be thorough but concise. If no filings found, say so clearly."""

    return run_subagent("Filing Search", system_prompt, task, FILING_SEARCH_TOOLS)


def run_risk_analysis(task: str) -> str:
    """Risk Analysis Subagent."""
    system_prompt = """You are a UCC lien risk analysis specialist. Your ONLY job is to
assess lien risk for entities. Use calculate_risk_score to get quantitative risk profiles.
If needed, use search_filings for additional context.

Focus on: risk score, risk level, number of liens, collateral breadth (blanket vs specific),
number of jurisdictions, secured party concentration. Always include the recommendation.
Cite specific numbers and filing details."""

    return run_subagent("Risk Analysis", system_prompt, task, RISK_ANALYSIS_TOOLS)


# =============================================================================
# COORDINATOR
# =============================================================================

class Coordinator:
    """Orchestrates the UCC Filing Research System with conversation memory."""

    def __init__(self):
        self.history: list[dict] = []

    def _get_history_context(self) -> str:
        if not self.history:
            return "No previous conversation."
        lines = ["## Previous Conversation"]
        for entry in self.history[-MAX_HISTORY_TURNS * 2:]:
            role = "User" if entry["role"] == "user" else "Assistant"
            content = entry["content"][:200]
            lines.append(f"**{role}**: {content}")
        return "\n".join(lines)

    def _classify_intent(self, query: str) -> str:
        """Determine which subagent(s) to invoke based on the query."""
        q = query.lower()
        needs_risk = any(w in q for w in ["risk", "exposure", "assess", "evaluate", "how risky", "risk level"])
        needs_search = any(w in q for w in ["find", "search", "filing", "filings", "look up", "what about"])

        if needs_risk and needs_search:
            return "both"
        elif needs_risk:
            return "risk"
        elif needs_search:
            return "search"
        else:
            # Default: do both for comprehensive research
            return "both"

    def run(self, user_query: str) -> str:
        """Process a user query through the coordinator."""
        observe("COORDINATOR", f"Received: {user_query}")

        # Add to history
        self.history.append({"role": "user", "content": user_query})
        history_context = self._get_history_context()

        # Classify intent
        intent = self._classify_intent(user_query)
        observe("COORDINATOR", f"Intent: {intent}")

        # Build task with conversation context
        task_with_context = f"""{user_query}

{history_context}"""

        # Execute subagent(s)
        filing_results = None
        risk_results = None

        if intent in ("search", "both"):
            observe("COORDINATOR", "Dispatching Filing Search subagent")
            filing_results = run_filing_search(task_with_context)
            observe_handoff("Filing Search", "Coordinator", len(filing_results))

        if intent in ("risk", "both"):
            # If we have filing results, pass them as context
            risk_task = task_with_context
            if filing_results:
                risk_task += f"\n\n## Filing Data (from search)\n{filing_results}"
            observe("COORDINATOR", "Dispatching Risk Analysis subagent")
            risk_results = run_risk_analysis(risk_task)
            observe_handoff("Risk Analysis", "Coordinator", len(risk_results))

        # Synthesize results
        observe("COORDINATOR", "Synthesizing results")
        synthesis_parts = [f"User query: {user_query}"]
        if filing_results:
            synthesis_parts.append(f"## Filing Search Results\n{filing_results}")
        if risk_results:
            synthesis_parts.append(f"## Risk Analysis Results\n{risk_results}")
        synthesis_parts.append(f"\n{history_context}")

        synthesis = client.messages.create(
            model=MODEL,
            max_tokens=4096,
            system="""You are a coordinator synthesizing research results about UCC filings.
Combine the subagent findings into a clear, comprehensive response.
Cite specific filing numbers and data. If referencing previous conversation, be explicit about what changed.""",
            messages=[{"role": "user", "content": "\n\n".join(synthesis_parts)}],
        )

        final = ""
        for block in synthesis.content:
            if hasattr(block, "text"):
                final += block.text

        # Add to history
        self.history.append({"role": "assistant", "content": final[:500]})

        observe("COORDINATOR", "Complete")
        return final


# =============================================================================
# MAIN
# =============================================================================

if __name__ == "__main__":
    print("=" * 60)
    print("M15B — Coordinator + Subagents (SOLUTION)")
    print("=" * 60)

    coord = Coordinator()

    print("\n\n>>> Turn 1: Find Acme filings in New York")
    r1 = coord.run("Find all UCC filings for Acme Corporation in New York")
    print(f"\nANSWER:\n{r1}")

    print("\n\n>>> Turn 2: Risk assessment")
    r2 = coord.run("What's the overall risk level for Acme Corporation?")
    print(f"\nANSWER:\n{r2}")

    print("\n\n>>> Turn 3: Follow-up about Texas")
    r3 = coord.run("What about their filings in Texas?")
    print(f"\nANSWER:\n{r3}")

    print("\n\n>>> Turn 4: Nonexistent entity")
    r4 = coord.run("Find filings for NonExistent Corp")
    print(f"\nANSWER:\n{r4}")
